Fix task api status code response for malformed id in the scheduler

jpbruinsslot opened the pull request:
### Changes

A 500 was returned on malformed id's when issuing requests to the `/tasks/{id}` endpoint. This PR solves this by correctly relaying a 400 status code error. I opted for catching exceptions from the `storage` module by implementing a `StorageError` exception, in order to keep the server API clean and not litter it with specific sqlalchemy, and psycopg2 exception handling.

### Issue link

Closes  #2943

ammar92 commented:
User requests with malformed input should indeed yield a `400` status response. However, I'm still a bit hesitant about throwing `400`s for every `StorageException`, since errors such as `External service returned an error.` doesn't necessarily indicate that the request is wrong. Perhaps within this context it's mostly a client error (due to a parse error to UUIDs?), but the error message is misleading.

I was also wondering; when should it be a not found error or a malformed id error? They're very identical, because with an invalid format you could still say that the entity cannot be found since it simply doesn't exit. But, if `task_id` is, for example, an expected UUID, then that could be validated. FastAPI supports a few extra data types, including UUIDs, that can be validated, even on on path operations. E.g.:
```python3
@app.put("/items/{item_id}")
def read_items(
    item_id: UUID,
    ...
):
```

Let me know what you think. I do like the framework you implemented to handle these errors.

jpbruinsslot commented:
I agree with the points that you've raised. I didn't realize that fastapi could leverage the fastapi validation for it that would make it more elegant.

jpbruinsslot commented:
@ammar92 I've updated the pr and add the fastapi uuid suggestion. I've restructured the `StorageError` a bit such that it is still a 500 when an exception occurs within the `storage` module. Additionally I updated the `FilterError` in the `filters` module, to make sure that those are specifically handled instead of a generic `StorageError`

ammar92 reviewed (approved):
Looks good to me 👍

jpbruinsslot commented:
Just fixing some pre-commit issues

stephanie0x00 commented:
## Checklist for QA:

- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [ ] I confirmed that there are no unintended functional regressions in this branch:
  - [x] I have managed to pass the onboarding flow
  - [x] Objects and Findings are created properly
  - [x] Tasks are created and completed properly
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.
- [ ] I checked the logs for errors and/or warnings and made issues where necessary

### What works:
- Onboarding flow works
- (Re)Scheduling of tasks work.
- Findings are created

### What doesn't work:
When reproducing the steps as described in #2943 I get the following error: 

![image](https://github.com/minvws/nl-kat-coordination/assets/9821756/277122b7-2c56-4e17-b82a-a95c86e7f1a6)


```
rocky-1  | [21/May/2024 09:15:41] "GET /en/aa/tasks/d42ccb45-8aa5-4d5d-bfbe-02f295c90df0/download/ HTTP/1.1" 200 494
rocky-1  | Internal Server Error: /en/aa/tasks/d42ccb45-8aa5-4d5d-bfbe-02f295c90ueue/download/
rocky-1  | Traceback (most recent call last):
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/django/core/handlers/exception.py", line 55, in inner
rocky-1  |     response = get_response(request)
rocky-1  |                ^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/django/core/handlers/base.py", line 197, in _get_response
rocky-1  |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
rocky-1  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/django/views/generic/base.py", line 104, in view
rocky-1  |     return self.dispatch(request, *args, **kwargs)
rocky-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/django/views/generic/base.py", line 143, in dispatch
rocky-1  |     return handler(request, *args, **kwargs)
rocky-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/app/rocky/rocky/views/tasks.py", line 28, in get
rocky-1  |     task_details = client.get_task_details(self.organization.code, task_id)
rocky-1  |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/app/rocky/rocky/scheduler.py", line 227, in get_task_details
rocky-1  |     res.raise_for_status()
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_models.py", line 761, in raise_for_status
rocky-1  |     raise HTTPStatusError(message, request=request, response=self)
rocky-1  | httpx.HTTPStatusError: Client error '422 Unprocessable Entity' for url 'http://scheduler:8000/tasks/d42ccb45-8aa5-4d5d-bfbe-02f295c90ueue'
rocky-1  | For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422
rocky-1  | Internal Server Error: /en/aa/tasks/d42ccb45-8aa5-4d5d-bfbe-02f295c90ueue/download/
rocky-1  | Traceback (most recent call last):
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/django/core/handlers/exception.py", line 55, in inner
rocky-1  |     response = get_response(request)
rocky-1  |                ^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/django/core/handlers/base.py", line 197, in _get_response
rocky-1  |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
rocky-1  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/django/views/generic/base.py", line 104, in view
rocky-1  |     return self.dispatch(request, *args, **kwargs)
rocky-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/django/views/generic/base.py", line 143, in dispatch
rocky-1  |     return handler(request, *args, **kwargs)
rocky-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/app/rocky/rocky/views/tasks.py", line 28, in get
rocky-1  |     task_details = client.get_task_details(self.organization.code, task_id)
rocky-1  |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/app/rocky/rocky/scheduler.py", line 227, in get_task_details
rocky-1  |     res.raise_for_status()
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_models.py", line 761, in raise_for_status
rocky-1  |     raise HTTPStatusError(message, request=request, response=self)
rocky-1  | httpx.HTTPStatusError: Client error '422 Unprocessable Entity' for url 'http://scheduler:8000/tasks/d42ccb45-8aa5-4d5d-bfbe-02f295c90ueue'
rocky-1  | For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422
rocky-1  | [21/May/2024 09:15:47] "GET /en/aa/tasks/d42ccb45-8aa5-4d5d-bfbe-02f295c90
```

### Bug or feature?:
See above.

Donnype commented on mula/scheduler/storage/errors.py:
```diff
@@ -0,0 +1,20 @@
+    pass
+
+
+def exception_handler(func):
```
```suggestion
def db_exception_handler(func):
```
?

Donnype commented on mula/scheduler/server/server.py:
```diff
@@ -398,12 +402,12 @@ def list_tasks(
-                status_code=fastapi.status.HTTP_400_BAD_REQUEST,
-                detail=str(exc),
+                status_code=fastapi.status.HTTP_500_INTERNAL_SERVER_ERROR,
+                detail=f"error occurred while accessing the database [exception: {exc}]",
```
We could consider a generic `@app.exception_handler` as well, which would clean up the router code but forces you to combing thinking about Exception hierarchy with a hierarchy of exception_handlers.

jpbruinsslot replied:
Yes I agree, let's pick it up in a new issue

Donnype commented on mula/tests/integration/test_api.py:
```diff
@@ -505,6 +595,16 @@ def test_get_task(self):
 
+    def test_get_task_malformed_id(self):
+        response = self.client.get("/tasks/123.123")
+        self.assertEqual(422, response.status_code)
```
:+1:

Donnype reviewed (approved):
:ok_hand: 

Check out the suggestions!

Rieven commented:
> ## Checklist for QA:
> * [x]  I have checked out this branch, and successfully ran a fresh `make reset`.
> * [ ]  I confirmed that there are no unintended functional regressions in this branch:
>   
>   * [x]  I have managed to pass the onboarding flow
>   * [x]  Objects and Findings are created properly
>   * [x]  Tasks are created and completed properly
> * [ ]  I confirmed that the PR's advertised `feature` or `hotfix` works as intended.
> * [ ]  I checked the logs for errors and/or warnings and made issues where necessary
> 
> ### What works:
> * Onboarding flow works
> * (Re)Scheduling of tasks work.
> * Findings are created
> 
> ### What doesn't work:
> When reproducing the steps as described in #2943 I get the following error:
> 
> ![image](https://private-user-images.githubusercontent.com/9821756/332368323-277122b7-2c56-4e17-b82a-a95c86e7f1a6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY0NjAwMTUsIm5iZiI6MTcxNjQ1OTcxNSwicGF0aCI6Ii85ODIxNzU2LzMzMjM2ODMyMy0yNzcxMjJiNy0yYzU2LTRlMTctYjgyYS1hOTVjODZlN2YxYTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MjNUMTAyMTU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Mjk5MTAxN2JlMmJkNTIzN2ZiNDFhMWJkNjc2NDJhMzIzNDQxZmE0ZDkzZTBiNjNmNmJmZDRlYTYwYjUxZDk0MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.POp-Dh9sZQF0LGrNKOpWPgG4hxlEPnvCoZ39aUZjKSE)

@stephanie0x00 This will be solved in this PR: https://github.com/minvws/nl-kat-coordination/pull/1957

jpbruinsslot commented:
For QA: since it is obscure to trigger this error, the 500 error is resolved in the scheduler, and a PR to handle the 422 status error code is being finalized, I would suggest to merge this so that the new rocky PR has the updated status code reporting from the scheduler.
